### ISSUE TITLE
Adding assumptions to variables to satisfy LH

### ISSUE THREAD

**Issue by lippirk** [ucsd-progsys/liquidhaskell#1452]
**tldr;** Is there a way to tell LH to make an assumption about a variable? 

---------------------------------------------------
For example:
```haskell
import Data.Vector as V

{-@ scannedVector :: xs:Vector a -> {ys:Vector a | vlen ys == 1 + vlen xs} @-} 
scannedVector :: Vector a -> Vector a
scannedVector xs = iscanl' (\idx _ _ -> xs V.! idx) 0 xs
```
LH quite rightly rejects this because it doesn't know that `0 <= idx < vlen xs`. 

If we try to write a spec for this:

```
iscanl' :: ({n:Nat | n < vlen xs} -> a -> b -> a)
       -> a
       -> xs:(Data.Vector.Vector b)
        -> {ys:Data.Vector.Vector a | 1 + vlen xs == vlen ys}
```
then it doesn't work; I assume this is because of currying, because rearranging the arguments _does_ work:

```
{-@ assume my_iscanl' :: xs:(Data.Vector.Vector b)
                                  -> (n:{n:Nat | n < vlen xs} -> a -> b -> a)
                                  -> a
                                  -> {ys:Data.Vector.Vector a | 1 + vlen xs == vlen ys}
  @-}
```

I'd rather not to have to rearrange arguments. I'd quite like to be able to do something like this:

```haskell
scannedVector :: Vector a -> Vector a
scannedVector xs = iscanl' (\idx _ _ -> 
    {-@ assume :: idx:{n:Nat | n < vlen xs} @-} xs V.! idx) 0 xs
```

but this doesn't work. I am wondering what you would recommend here?

PS: thank you for helping me get liquidhaskell to work these past few days, I'm really grateful

**Comment by ranjitjhala** [comment]
Hi @lippirk -- great question! You can either use `liquidAssume` defined as below:

```haskell
import qualified Data.Vector as V

-- or you can do
-- import Language.Haskell.Liquid.Prelude (liquidAssume)

{-@ liquidAssume :: b:Bool -> a -> {v: a | b}  @-}
liquidAssume :: Bool -> a -> a
liquidAssume b x = if b then x else undefined

scannedVector :: (Num a) => V.Vector a -> V.Vector a
scannedVector xs = V.iscanl' (\idx _ _ -> (liquidAssume (0 <= idx && idx < V.length xs) xs) V.! idx) 0 xs
```

or there is a more fancy way of writing a signature for `iscanl'` that 
allows you to NOT have to flip the arguments, will post that in a little bit 
after my lecture...

**Comment by lippirk** [comment]
Cool! That looks like exactly what I need. Thank you.

Please do, I'd be really interested seeing how you manage to refine `iscanl'` like that.

I haven't been looking at the haddocks for LH - only googling - but it looks like I ought to be. Hopefully this will be the last time I bother you

**Comment by ranjitjhala** [comment]
Hi @lippirk -- sorry for the delay, but here's a worked out example that illustrates how to use 
[bounded refinement types](https://ranjitjhala.github.io/static/bounded_refinement_types.pdf) to -- allow you to write a suitable signature for `iscanl'`. Its a bit tricky 
I admit :-( but still its possible! 

PS: Please don't hesitate to ask us Qs on github or slack or mail -- we _welcome_ it, as its the best 
way for us to find out what things need to be explained better, require more examples and so on!

I am adding this as an example in `tests/absref/{pos,neg}/FlipArgs.hs`

```haskell
-------------------------------------------------------------------------------
-- | How to use bounds when the argument orders are backwards -----------------
--   https://github.com/ucsd-progsys/liquidhaskell/issues/1452
-------------------------------------------------------------------------------

import qualified Data.Vector as V 

-------------------------------------------------------------------------------
-- | Warmup example: suppose we require `bar` to have the following type
-------------------------------------------------------------------------------

{-@ bar :: x:Int -> {y:Int | y < x} -> Int @-}
bar :: Int -> Int -> Int 
bar x y = x + y 

-------------------------------------------------------------------------------
-- | `testOK` is SAFE as the signature of `fooOK` says it only calls its 
--    argument with values less than `a` 
-------------------------------------------------------------------------------

{-@ fooOK :: a:Int -> ({v:Int | v <= a} -> Int) -> Int @-} 
fooOK :: Int -> (Int -> Int) -> Int 
fooOK a f = f a + f (a - 1) + f (a - 2)

testOK :: Int
testOK = fooOK 7 (bar 10) 


-------------------------------------------------------------------------------
{- | Naively, `testBAD` is REJECTED as we have no way to say what values the 
     closure will be called with! (Remove the signature to see this)
     However in the below, we say that the closure is called with Int-satisfying `p`
     where the second Int satisfies `q` SUCH THAT

       forall n, xq. (q xq) => (n <= xq) => (p n) 

     that is only values LESS THAN `xq` satisfy `p`.
     This allows LH to deduce at the callsite 
   
       fooBAD (bar 10) 7
       
     that `bar 10` is only called with Int values that are `<= 7` and hence is safe
     w.r.t. the specification for `bar`!
 -}

{-@ fooBAD :: forall <p :: Int -> Bool, q :: Int -> Bool>. 
                { a :: Int<q> |- {v:Int | v <= a} <: Int<p> }
                (Int<p> -> Int) -> Int<q> -> Int
  @-}
fooBAD :: (Int -> Int) -> Int -> Int 
fooBAD f a = f a + f (a - 1) + f (a - 2)

testBAD :: Int
testBAD = fooBAD (bar 10) 7 

-------------------------------------------------------------------------------
{- | In the below signature, the "bound" says that the scanner will be called
     with indices of type `Int<p>` when run on a `q`-refined vector only when:
     `p` and `q` satisfy a particular relationship, namely:
      
        forall v, vec. (q vec) => 0 <= n < vlen v => (p n)
      
     That is, any `Int` satisfying `p` is indeed a valid index for a `vec` satisfyin `q`.
   
     As the actual index satisfies `p` and the vector `xs` satisfies `q`, LH 
     deduces that the indices used at the call-site within `scannedVector` 
     are valid for `xs`.
 -}    

{-@ assume V.iscanl' :: forall <p :: Int -> Bool, q :: V.Vector b -> Bool>.
      { vec :: V.Vector b <<q>> |- {n:Int | 0 <= n && n < vlen vec} <: Int<p> }
      (Int<p> -> a -> b -> a)
      -> a 
      -> xs:(V.Vector b <<q>>) 
      -> {ys:V.Vector a | 1 + vlen xs == vlen ys}
  @-}

scannedVector :: (Num a) => V.Vector a -> V.Vector a
scannedVector xs = V.iscanl' (\idx _ _ -> xs V.! idx) 0 xs
```

/cc @nikivazou -- you may like the above application of bounds.

**Comment by lippirk** [comment]
Definitely would not have come up with that! But seems to make sense after staring at it for a while. Thank you!

Out of interest is there a reason why there aren't many vector specs (compared to the number of functions exported by `Data.Vector`)? I added a few refinements type for the functions that I wanted to use - it would be nice to merge these in eventually when I get round to testing them a bit more (and after utilizing the above technique!)
